File: src/entities.js
// Pickup + enemy entities. Slime wanders; chases when player is close.

const DIR_VEC = {
  down:  { x: 0,  y: 1 },
  up:    { x: 0,  y: -1 },
  left:  { x: -1, y: 0 },
  right: { x: 1,  y: 0 },
};

const BOLT_FW = 24, BOLT_FH = 20;
const IMPACT_FW = 32, IMPACT_FH = 32;
const BURN_FW = 14, BURN_FH = 14;
const IMPACT_DUR = 0.42;
const BURN_DUR = 1.6;

// Rotation angle so the projectile's "forward" (positive x in sheet) points
// in the requested direction.
const DIR_ANGLE = {
  right: 0,
  down:  Math.PI / 2,
  left:  Math.PI,
  up:    -Math.PI / 2,
};

export class MagicBolt {
  constructor(player, assets) {
    this.kind = "bolt";
    this.facing = player.facing;
    this.w = 8;
    this.h = 8;
    const o = player.castOrigin();
    const d = DIR_VEC[player.facing];
    this.x = o.x - this.w / 2 + d.x * 4;
    this.y = o.y - this.h / 2 + d.y * 4;
    this.vx = d.x * 140;
    this.vy = d.y * 140;
    this.life = 0.9;
    this.t = 0;
    this.dead = false;
    this.sheet = assets?.images?.["fx_projectile.png"] || null;
  }
  update(dt, _player, map) {
    this.t += dt;
    this.life -= dt;
    if (this.life <= 0) { this.dead = true; return; }
    const nx = this.x + this.vx * dt;
    const ny = this.y + this.vy * dt;
    if (map.rectSolid(nx, ny, this.w, this.h)) { this.dead = true; return; }
    this.x = nx;
    this.y = ny;
  }
  draw(ctx, camera) {
    const cx = Math.round(this.x + this.w / 2 - camera.x);
    const cy = Math.round(this.y + this.h / 2 - camera.y);
    if (this.sheet) {
      const frame = ((this.t * 14) | 0) % 4;
      ctx.save();
      ctx.translate(cx, cy);
      ctx.rotate(DIR_ANGLE[this.facing] ?? 0);
      ctx.drawImage(this.sheet, frame * BOLT_FW, 0, BOLT_FW, BOLT_FH,
                    -BOLT_FW / 2, -BOLT_FH / 2, BOLT_FW, BOLT_FH);
      ctx.restore();
      return;
    }
    // procedural fallback
    const flick = ((this.t * 20) | 0) % 2;
    ctx.fillStyle = "rgba(255,220,120,0.35)";
    ctx.fillRect(cx - 5, cy - 5, 10, 10);
    ctx.fillStyle = flick ? "#fff7c4" : "#fde36a";
    ctx.fillRect(cx - 2, cy - 2, 4, 4);
  }
}

export class Impact {
  constructor(x, y, assets) {
    this.kind = "impact";
    this.x = x - IMPACT_FW / 2;
    this.y = y - IMPACT_FH / 2;
    this.w = 16;
    this.h = 16;
    this.t = 0;
    this.dead = false;
    this.sheet = assets?.images?.["fx_impact.png"] || null;
    this.damaged = new Set(); // slimes already damaged by this burst
  }
  update(dt) {
    this.t += dt;
    if (this.t >= IMPACT_DUR) this.dead = true;
  }
  // small AOE damage in the first 1/3 of the impact
  damageRect() {
    if (this.t > IMPACT_DUR * 0.4) return null;
    return { x: this.x + 4, y: this.y + 4, w: IMPACT_FW - 8, h: IMPACT_FH - 8 };
  }
  draw(ctx, camera) {
    if (!this.sheet) return;
    const p = Math.min(0.999, this.t / IMPACT_DUR);
    const frame = Math.floor(p * 6);
    const dx = Math.round(this.x - camera.x);
    const dy = Math.round(this.y - camera.y);
    ctx.drawImage(this.sheet, frame * IMPACT_FW, 0, IMPACT_FW, IMPACT_FH,
                  dx, dy, IMPACT_FW, IMPACT_FH);
  }
}

export class Burning {
  constructor(x, y, assets) {
    this.kind = "burning";
    this.x = x - BURN_FW / 2;
    this.y = y - BURN_FH / 2 + 4;
    this.w = BURN_FW;
    this.h = BURN_FH - 4;
    this.t = 0;
    this.dead = false;
    this.sheet = assets?.images?.["fx_burn.png"] || null;
    this.tickAccum = 0;
  }
  update(dt) {
    this.t += dt;
    this.tickAccum += dt;
    if (this.t >= BURN_DUR) this.dead = true;
  }
  // tick once every 0.3s for DoT
  consumeTick() {
    if (this.tickAccum >= 0.3) {
      this.tickAccum = 0;
      return true;
    }
    return false;
  }
  draw(ctx, camera) {
    const dx = Math.round(this.x - camera.x);
    const dy = Math.round(this.y - camera.y);
    if (this.sheet) {
      const frame = ((this.t * 10) | 0) % 4;
      const fadeOut = this.t > BURN_DUR - 0.3;
      if (fadeOut) ctx.globalAlpha = Math.max(0, (BURN_DUR - this.t) / 0.3);
      ctx.drawImage(this.sheet, frame * BURN_FW, 0, BURN_FW, BURN_FH,
                    dx, dy, BURN_FW, BURN_FH);
      ctx.globalAlpha = 1;
    } else {
      ctx.fillStyle = "#e54b4b";
      ctx.fillRect(dx + 4, dy + 4, 6, 6);
    }
  }
}
export class Coin {
  constructor(x, y) {
    this.kind = "coin";
    this.x = x;
    this.y = y;
    this.w = 6;
    this.h = 6;
    this.bob = Math.random() * Math.PI * 2;
    this.dead = false;
  }
  update(dt) {
    this.bob += dt * 5;
  }
  draw(ctx, camera) {
    const dy = Math.sin(this.bob) * 1.5;
    const px = Math.round(this.x - camera.x);
    const py = Math.round(this.y - camera.y + dy);
    ctx.fillStyle = "rgba(0,0,0,0.3)";
    ctx.fillRect(px, this.y - camera.y + 6 | 0, 6, 1);
    ctx.fillStyle = "#caa53b";
    ctx.fillRect(px, py, 6, 6);
    ctx.fillStyle = "#fde36a";
    ctx.fillRect(px + 1, py + 1, 4, 4);
    ctx.fillStyle = "#fff7c4";
    ctx.fillRect(px + 2, py + 1, 1, 1);
  }
}

export class Slime {
  constructor(x, y) {
    this.kind = "slime";
    this.x = x;
    this.y = y;
    this.w = 12;
    this.h = 10;
    this.vx = 0;
    this.vy = 0;
    this.tick = Math.random() * 2;
    this.dirTimer = 0;
    this.dead = false;
    this.squish = 0;
  }
  update(dt, player, map) {
    this.tick += dt;
    this.squish = (Math.sin(this.tick * 4) + 1) * 0.5;

    const dx = (player.x + player.w / 2) - (this.x + this.w / 2);
    const dy = (player.y + player.h / 2) - (this.y + this.h / 2);
    const dist2 = dx * dx + dy * dy;

    if (dist2 < 64 * 64) {
      // chase
      const inv = 1 / Math.sqrt(dist2 || 1);
      this.vx = dx * inv * 30;
      this.vy = dy * inv * 30;
    } else {
      this.dirTimer -= dt;
      if (this.dirTimer <= 0) {
        const a = Math.random() * Math.PI * 2;
        this.vx = Math.cos(a) * 18;
        this.vy = Math.sin(a) * 18;
        this.dirTimer = 1.0 + Math.random() * 1.5;
      }
    }

    const mx = this.vx * dt;
    const my = this.vy * dt;
    if (!map.rectSolid(this.x + mx, this.y, this.w, this.h)) this.x += mx;
    else this.vx *= -1;
    if (!map.rectSolid(this.x, this.y + my, this.w, this.h)) this.y += my;
    else this.vy *= -1;
  }
  draw(ctx, camera) {
    const px = Math.round(this.x - camera.x);
    const py = Math.round(this.y - camera.y);
    const sq = this.squish;
    ctx.fillStyle = "rgba(0,0,0,0.35)";
    ctx.fillRect(px + 1, py + this.h - 1, this.w - 2, 2);
    ctx.fillStyle = "#4ab07a";
    ctx.fillRect(px + 1, py + 3 - (sq | 0), 10, 7 + (sq | 0));
    ctx.fillStyle = "#7ad8a6";
    ctx.fillRect(px + 2, py + 3 - (sq | 0), 8, 1);
    ctx.fillStyle = "#fff";
    ctx.fillRect(px + 3, py + 5, 2, 2);
    ctx.fillRect(px + 7, py + 5, 2, 2);
    ctx.fillStyle = "#000";
    ctx.fillRect(px + 4, py + 6, 1, 1);
    ctx.fillRect(px + 8, py + 6, 1, 1);
  }
}

File: python/entities.py
import math
import random
import pygame

DIR_VEC = {
    "down":  (0, 1),
    "up":    (0, -1),
    "left":  (-1, 0),
    "right": (1, 0),
}

BOLT_FW, BOLT_FH = 24, 20
IMPACT_FW, IMPACT_FH = 32, 32
BURN_FW, BURN_FH = 14, 14
IMPACT_DUR = 0.42
BURN_DUR = 1.6

# Degrees (pygame.transform.rotate uses degrees, counter-clockwise positive).
# The projectile sheet points right; rotating CCW by these values aims it.
DIR_DEG = {
    "right": 0,
    "down":  -90,
    "left":  180,
    "up":    90,
}


class MagicBolt:
    kind = "bolt"

    def __init__(self, player, assets):
        self.facing = player.facing
        d = DIR_VEC[player.facing]
        self.x = player.x + player.w / 2 - 4 + d[0] * 8
        self.y = player.y + player.h / 2 - 4 + d[1] * 8
        self.w = 8
        self.h = 8
        self.vx = d[0] * 140
        self.vy = d[1] * 140
        self.life = 0.9
        self.t = 0.0
        self.dead = False
        sheet = assets["images"].get("fx_projectile.png") if assets else None
        self.frames = None
        if sheet is not None:
            self.frames = []
            for i in range(4):
                sub = sheet.subsurface((i * BOLT_FW, 0, BOLT_FW, BOLT_FH))
                self.frames.append(pygame.transform.rotate(sub, DIR_DEG[self.facing]))

    def update(self, dt, _player, tmap):
        self.t += dt
        self.life -= dt
        if self.life <= 0:
            self.dead = True
            return
        nx = self.x + self.vx * dt
        ny = self.y + self.vy * dt
        if tmap.rect_solid(nx, ny, self.w, self.h):
            self.dead = True
            return
        self.x = nx
        self.y = ny

    def draw(self, surf, camera):
        cx = round(self.x + self.w / 2 - camera.x)
        cy = round(self.y + self.h / 2 - camera.y)
        if self.frames:
            frame = int(self.t * 14) % 4
            img = self.frames[frame]
            r = img.get_rect(center=(cx, cy))
            surf.blit(img, r)
            return
        # fallback
        surf.fill((255, 220, 120), (cx - 5, cy - 5, 10, 10))
        col = (255, 247, 196) if int(self.t * 20) % 2 else (253, 227, 106)
        surf.fill(col, (cx - 2, cy - 2, 4, 4))


class Impact:
    kind = "impact"

    def __init__(self, x, y, assets):
        self.x = x - IMPACT_FW / 2
        self.y = y - IMPACT_FH / 2
        self.w = 16
        self.h = 16
        self.t = 0.0
        self.dead = False
        self.sheet = assets["images"].get("fx_impact.png") if assets else None
        self.damaged = set()

    def update(self, dt, _player=None, _map=None):
        self.t += dt
        if self.t >= IMPACT_DUR:
            self.dead = True

    def damage_rect(self):
        if self.t > IMPACT_DUR * 0.4:
            return None
        return (self.x + 4, self.y + 4, IMPACT_FW - 8, IMPACT_FH - 8)

    def draw(self, surf, camera):
        if not self.sheet:
            return
        p = min(0.999, self.t / IMPACT_DUR)
        frame = int(p * 6)
        dx = round(self.x - camera.x)
        dy = round(self.y - camera.y)
        surf.blit(self.sheet, (dx, dy), (frame * IMPACT_FW, 0, IMPACT_FW, IMPACT_FH))


class Burning:
    kind = "burning"

    def __init__(self, x, y, assets):
        self.x = x - BURN_FW / 2
        self.y = y - BURN_FH / 2 + 4
        self.w = BURN_FW
        self.h = BURN_FH - 4
        self.t = 0.0
        self.dead = False
        self.sheet = assets["images"].get("fx_burn.png") if assets else None
        self.tick_accum = 0.0

    def update(self, dt, _player=None, _map=None):
        self.t += dt
        self.tick_accum += dt
        if self.t >= BURN_DUR:
            self.dead = True

    def consume_tick(self):
        if self.tick_accum >= 0.3:
            self.tick_accum = 0.0
            return True
        return False

    def draw(self, surf, camera):
        dx = round(self.x - camera.x)
        dy = round(self.y - camera.y)
        if self.sheet:
            frame = int(self.t * 10) % 4
            sub = self.sheet.subsurface((frame * BURN_FW, 0, BURN_FW, BURN_FH))
            if self.t > BURN_DUR - 0.3:
                a = max(0.0, (BURN_DUR - self.t) / 0.3)
                tmp = sub.copy()
                tmp.set_alpha(int(255 * a))
                surf.blit(tmp, (dx, dy))
            else:
                surf.blit(sub, (dx, dy))
        else:
            surf.fill((229, 75, 75), (dx + 4, dy + 4, 6, 6))


class Coin:
    kind = "coin"

    def __init__(self, x, y, assets=None):
        self.x = x
        self.y = y
        self.w = 6
        self.h = 6
        self.bob = random.random() * math.pi * 2
        self.dead = False

    def update(self, dt, _player=None, _map=None):
        self.bob += dt * 5

    def draw(self, surf, camera):
        dy = math.sin(self.bob) * 1.5
        px = round(self.x - camera.x)
        py = round(self.y - camera.y + dy)
        surf.fill((40, 30, 0), (px, int(self.y - camera.y + 6), 6, 1))
        surf.fill((202, 165, 59), (px, py, 6, 6))
        surf.fill((253, 227, 106), (px + 1, py + 1, 4, 4))
        surf.fill((255, 247, 196), (px + 2, py + 1, 1, 1))


class Slime:
    kind = "slime"

    def __init__(self, x, y, assets=None):
        self.x = x
        self.y = y
        self.w = 12
        self.h = 10
        self.vx = 0.0
        self.vy = 0.0
        self.tick = random.random() * 2
        self.dir_timer = 0.0
        self.dead = False
        self.squish = 0.0

    def update(self, dt, player, tmap):
        self.tick += dt
        self.squish = (math.sin(self.tick * 4) + 1) * 0.5

        dx = (player.x + player.w / 2) - (self.x + self.w / 2)
        dy = (player.y + player.h / 2) - (self.y + self.h / 2)
        dist2 = dx * dx + dy * dy

        if dist2 < 64 * 64:
            inv = 1.0 / math.sqrt(dist2 or 1)
            self.vx = dx * inv * 30
            self.vy = dy * inv * 30
        else:
            self.dir_timer -= dt
            if self.dir_timer <= 0:
                a = random.random() * math.pi * 2
                self.vx = math.cos(a) * 18
                self.vy = math.sin(a) * 18
                self.dir_timer = 1.0 + random.random() * 1.5

        mx = self.vx * dt
        my = self.vy * dt
        if not tmap.rect_solid(self.x + mx, self.y, self.w, self.h):
            self.x += mx
        else:
            self.vx *= -1
        if not tmap.rect_solid(self.x, self.y + my, self.w, self.h):
            self.y += my
        else:
            self.vy *= -1

    def draw(self, surf, camera):
        px = round(self.x - camera.x)
        py = round(self.y - camera.y)
        sq = int(self.squish)
        shadow = pygame.Surface((self.w - 2, 2), pygame.SRCALPHA)
        shadow.fill((0, 0, 0, 90))
        surf.blit(shadow, (px + 1, py + self.h - 1))
        surf.fill((74, 176, 122), (px + 1, py + 3 - sq, 10, 7 + sq))
        surf.fill((122, 216, 166), (px + 2, py + 3 - sq, 8, 1))
        surf.fill((255, 255, 255), (px + 3, py + 5, 2, 2))
        surf.fill((255, 255, 255), (px + 7, py + 5, 2, 2))
        surf.fill((0, 0, 0), (px + 4, py + 6, 1, 1))
        surf.fill((0, 0, 0), (px + 8, py + 6, 1, 1))
